### PR TITLE
SI-6932 StackOverflowError in chained Future.flatMap calls

### PR DESCRIPTION
**Pull request by phaller**
After discussion on commit 089805510ee79c389ff2a870355c8496c723b8cc, submitting a self-contained PR for the bug fix taken from:
https://github.com/viktorklang/scala/pull/5/commits

I agree with the assessment of @jsuereth, @retronym, and @adriaanm that package-private visibility should enable these changes for 2.10.x.

Note: already reviewed and discussed, but maybe we can have another completing review by @heathermiller

### CODE REVIEW

**Comment by phaller** [comment]
After this has been merged into 2.10.x I'm going to backport to 2.9.x.

File: src/library/scala/concurrent/BatchingExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    override def run(): Unit = {
+      require(_tasksLocal.get eq null)
+
+      val prevBlockContext = BlockContext.current
+      BlockContext.withBlockContext(this) {
+        try {
+          parentBlockContext = prevBlockContext
+
+          @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
+            case Nil ⇒ ()
```
Comment (heathermiller) [inline src/library/scala/concurrent/BatchingExecutor.scala]: This is a tiny nitpick, but don't we typically avoid unicode arrows in the Scala codebase? I seldom see them.

File: src/library/scala/concurrent/BatchingExecutor.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+package scala.concurrent
+
+import java.util.concurrent.{ Executor }
+import scala.concurrent._
+import scala.annotation.tailrec
+
+/**
+ * All Batchables are automatically batched when submitted to a BatchingExecutor
+ */
+private[concurrent] trait Batchable extends Runnable {
```
Comment (heathermiller) [inline src/library/scala/concurrent/BatchingExecutor.scala]: Shouldn't we remove the `Batchable` trait? It's unused, no?

Comment (phaller) [inline src/library/scala/concurrent/BatchingExecutor.scala]: It's indeed unused. Removed in new commit.

File: src/library/scala/concurrent/BatchingExecutor.scala (hunk 1 above)
Comment (phaller) [inline src/library/scala/concurrent/BatchingExecutor.scala]: Yeah, I replaced them.

**Comment by heathermiller** [comment]
Thanks for addressing! LGTM!

**Pull request merged**